Tests/Realm.Tests/Sync/DataTypeTests.cs

```diff
@@ -570,6 +569,65 @@ private static RealmValue Clone(RealmValue original)
             return clone;
         }
 
+        private static T CloneOrLookup<T>(T value, Realm targetRealm)
+        {
+            // If embedded - we need to clone as it might already be assigned to a different property
+            if (value is EmbeddedObject eobj)
+            {
+                return Operator.Convert<RealmObjectBase, T>(Clone(eobj).AsRealmObject());
+            }
+
+            // If RealmObject - we need to look up the existing equivalent in the correct realm
+            if (value is RealmObject robj)
+            {
+                // item2 belongs to realm2 - we want to look up the equivalent in realm1 to add it to dict1
+                Assert.That(robj.ObjectMetadata.Helper.TryGetPrimaryKeyValue(robj, out var pk), Is.True);
+                var item2InRealm1 = targetRealm.DynamicApi.FindCore(robj.ObjectSchema.Name, Operator.Convert<RealmValue>(pk));
+                return Operator.Convert<RealmObject, T>(item2InRealm1);
+            }
+
+            // If RealmValue that is holding an object, call CloneOrLookup
+            if (value is RealmValue rvalue && rvalue.Type == RealmValueType.Object)
+            {
+                var cloned = CloneOrLookup(rvalue.AsRealmObject(), targetRealm);
+                return Operator.Convert<RealmObjectBase, T>(cloned);
+            }
+
+            return value;
+        }
+
+        private static bool RealmValueEquals(RealmValue a, RealmValue b)
+        {
+            if (a.Equals(b))
+            {
+                return true;
+            }
+
+            // Special handling the object case as they might be "equivalent" but belonging to different realms
+            if (a.Type != RealmValueType.Object || b.Type != RealmValueType.Object)
+            {
+                return false;
+            }
+
+            var objA = a.AsRealmObject();
+            var objB = b.AsRealmObject();
+
+            if (objA.GetType() != objB.GetType())
```

Comment:
How is this different from line `607`? Don't we already know they are both `RealmValueType.Object`?

Reply:
This is not as obvious in the github diff view as it would have been in VS, but 607 is checking `RealmValue.Type` which is an enum of `int`, `string`, `object`, etc.

Line 617 is using reflection to get the .NET type of the `RealmObject` instance that we obtained. The [System.Type](https://docs.microsoft.com/en-us/dotnet/api/system.type?view=net-5.0) being returned here contains information about the actual .NET type - e.g. the class `IntPropertyObject` in the namespace `Realms.Tests` from the assembly XYZ, etc.